test/intl402/DisplayNames/prototype/of/type-language-invalid.js

```diff
@@ -0,0 +1,88 @@
+// Copyright (C) 2023 Igalia S.L. All rights reserved.
+// This code is governed by the BSD license found in the LICENSE file.
+/*---
+esid: sec-Intl.DisplayNames.prototype.of
+description: Throws a RangeError for invalid `language` codes
+info: |
+  12.3.3 Intl.DisplayNames.prototype.of ( code )
+
+  1. If type is "language", then
+      a. If code cannot be matched by the unicode_language_id Unicode locale nonterminal, throw a RangeError exception.
+      b. If IsStructurallyValidLanguageTag(code) is false, throw a RangeError exception.
+      c. Return CanonicalizeUnicodeLocaleId(code).
+features: [Intl.DisplayNames-v2]
+---*/
+
+var displayNames = new Intl.DisplayNames(undefined, {type: 'language'});
+
+assert.throws(RangeError, function() {
+  displayNames.of('');
+}, 'invalid language subtag - empty string');
+
+assert.throws(RangeError, function() {
+  displayNames.of('a');
+}, 'invalid language subtag - only one character');
+
+assert.throws(RangeError, function() {
+  displayNames.of('abcdefghi');
+}, 'invalid language subtag - > 8 characters');
+
+assert.throws(RangeError, function() {
+  displayNames.of('en-u-hebrew');
+}, 'invalid language id -- contains singleton');
```

Comment:
Can you elaborate on why you think this is invalid? It looks to me like a valid [Unicode locale identifier](https://www.unicode.org/reports/tr35/#Unicode_locale_identifier) with `unicode_language_id` "en" and Unicode locale extension sequence "-u-hebrew", which includes the `attribute` "hebrew" and no `keyword`s.

Reply:
When *type* is "language", *code* has to be a valid `unicode_language_id`. Although the string in question is a valid [`unicode_locale_id`](https://www.unicode.org/reports/tr35/#Unicode_locale_identifier), a [`unicode_language_id`](https://unicode.org/reports/tr35/#unicode_language_id) can only contain `unicode_language_subtag`s, `unicode_script_subtag`s, `unicode_region_subtag`s, and `unicode_variant_subtags`, so strings with singletons are necessarily invalid.

Reply:
Though now that I'm looking at it: doesn't the test in 12.5.1.1.a  duplicate the test in [IsStructurallyValidLanguageTag](https://tc39.es/ecma402/#sec-isstructurallyvalidlanguagetag)? Seems like all it does is save having to do the `ASCII-lowercase` at the start of IsStructurallyValidLanguageTag.

Reply:
> When _type_ is "language", _code_ has to be a valid `unicode_language_id`. See [12.5.1](https://tc39.es/ecma402/#sec-canonicalcodefordisplaynames). Although the string in question is a valid [`unicode_locale_id`](https://www.unicode.org/reports/tr35/#Unicode_locale_identifier), a [`unicode_language_id`](https://unicode.org/reports/tr35/#unicode_language_id) can only contain `unicode_language_subtag`, `unicode_script_subtag`, `unicode_region_subtag`, and `unicode_variant_subtag`, so strings with singletons are necessarily invalid.

:+1:

> Though now that I'm looking at it: doesn't the test in 12.5.1.1.a duplicate the test in [IsStructurallyValidLanguageTag](https://tc39.es/ecma402/#sec-isstructurallyvalidlanguagetag)? Seems like all it does is save having to do the `ASCII-lowercase` at the start of IsStructurallyValidLanguageTag.

Let's see... [CanonicalCodeForDisplayNames](https://tc39.es/ecma402/#sec-canonicalcodefordisplaynames) step 1.a ensures that the input is an isolated Unicode language identifier (_i.e., a language subtag with optional script subtag or a script subtag, in either case optionally followed by a region subtag and/or any number of variant subtags but NOT by singleton subtags or any other subtags_), and step 1.b ensures that it does not use backwards compatibility syntax (_[IsStructurallyValidLanguageTag](https://tc39.es/ecma402/#sec-isstructurallyvalidlanguagetag) step 3 rejecting e.g. "en_GB" and "root" and "Latn-DE" per the respective [UTS #35 Part 1 Core, BCP 47 Conformance](https://unicode.org/reports/tr35/#BCP_47_Conformance) "backwards compatibility" bullet points_) or contain duplicate variant subtags (_IsStructurallyValidLanguageTag step 6 rejecting e.g. "de-1996-fonipa-1996"_) that would both pass the preceding syntax check. So I like the current state of ECMA-402 text, and also the coverage being added in this PR.